core/ipc/twin/common/src/main/java/org/opennms/core/ipc/twin/common/AbstractTwinSubscriber.java

```diff
@@ -230,72 +136,173 @@ protected TwinUpdate mapTwinResponseToProto(byte[] responseBytes) {
+                      final int version,
+                      final JsonNode value) {
+            this.sessionId = Objects.requireNonNull(sessionId);
+            this.version = version;
+            this.value = Objects.requireNonNull(value);
         }
-        return null;
     }
 
-    private class SessionImpl<T> implements Closeable {
-
+    private class Subscription {
         private final String key;
-        private final Consumer<T> consumer;
-        private final Class<T> clazz;
 
-        public SessionImpl(String key, Class<T> clazz, Consumer<T> consumer) {
-            this.key = key;
-            this.clazz = clazz;
-            this.consumer = consumer;
-        }
+        private final Set<Consumer<JsonNode>> consumers = Sets.newConcurrentHashSet();
+
+        private Value value = null;
 
-        @Override
-        public void close() throws IOException {
-            sessionMap.remove(key, this);
-            LOG.info("Closed session with key {} ", key);
+        /**
+         * A tasked scheduled to retry an outstanding request.
+         *
+         * This is a one-shot task re-executing the request. The task will be scheduled when a request is emitted and
+         * will be canceled when an update is accepted. Setting this to {@code null} shows that there is no outstanding
+         * request.
+         */
+        private ScheduledFuture<?> retry = null;
+
+        private Subscription(final String key) {
+            this.key = Objects.requireNonNull(key);
         }
 
-        public void accept(byte[] objValue) throws IOException {
-            final T value = objectMapper.readValue(objValue, clazz);
-            LOG.trace("Updated consumer with key {}", key);
-            consumer.accept(value);
+        /**
+         * Consume the subscription.
+         *
+         * Adds a consumer to the subscription. Incoming values will be forwarded to the passed consumer until the
+         * returned value is closed. If there is a value already available for the subscription, the consumer will be
+         * called with this value immediately.
+         *
+         * @param clazz The class of the value to consume
+         * @param consumer The consumer accepting the values
+         * @param <T> The class of the value to consume
+         * @return a Closable, stopping the consumption when closed
+         */
+        public synchronized <T> Closeable consume(final Class<T> clazz, final Consumer<T> consumer) {
+            final Consumer<JsonNode> jsonConsumer = (json) -> {
+                try {
+                    // Deserialize to the final class
+                    final var value = AbstractTwinSubscriber.this.objectMapper.treeToValue(json, clazz);
+
+                    // Forward to typed consumer
+                    consumer.accept(value);
+
+                } catch (final JsonProcessingException e) {
+                    LOG.error("Deserialize twin failed: {} as {}", this.key, clazz, e);
+                }
```

Comment:
Admittedly, the `JsonProcessingException` is unlikely to happen. However, if it happens shouldn't it trigger a request for a new value? If `treeToValue` is called by the subscriber then it could catch the `JsonProcessingException` and either call `request` or throw an `IOException` that also would trigger a `request`.

Reply:
I would argue that en error in desertification is the subscribers fault. Even more, the consumer used to subscribe can throw an exception. Therefore I altered the exception handler to be more broad.

Forwarding the exception would mean to cancel all other consumers. Therefor logging is the only thing we can do there.